client/src/main/java/io/pravega/client/stream/Transaction.java

```diff
@@ -70,9 +70,11 @@
      * Send a transaction heartbeat and increase transaction's timeout by lease amount of milliseconds.
      *
      * @param lease Additional amount of time in milliseconds by which to increase transaction's timeout.
+     * @param switchOver Boolean indicating whether the previous successful ping request was performed on a different
+     *                   controller process by the client.
      * @throws PingFailedException Ping failed.
      */
-    void ping(long lease) throws PingFailedException;
+    void ping(long lease, boolean switchOver) throws PingFailedException;
```

Comment:
@bhargav-gulavani affinity is not implemented in the controller client. It can be done as an optimization but don't think its something implementations should depend on or explicitly exposed to the client. We are relying on controller's in-memory state for this feature? Also is there a way to make the switch over implicit in the server (something like - if no inmem state is found then treat it as a swithover scenario)

If we still need strict affinity implemented we need to be aware of a few things:
* Controller may not be truly stateless and will affect deployment, example we will need to explicitly expose each controller to the clients (today we are using mesos-dns transparently for this and could use a generic load-balancer in other deployments which has to account for this)
* Clients may not detect correctly when server failure has happened (distinguishing temporary network failures from server failures) and switchover will happen aggressively (should be uncommon though).